### DISCUSSION TITLE
Reducing out-of-context queries to the LLM

### DISCUSSION THREAD

**Discussion by wissamjur** [community/community#67504] [Programming Help] [answered]
Is reducing/eliminating out-of-context questions to the LLM done solely through the prompt input instructions?
I've had some success optimizing the prompt, but it occasionally veers off its given context.

**Comment by mantasu** [comment, marked as answer]
If queries are out of context, then LLM responses will also be out of context.  Here are a few tips besides **prompt-engineering**:

1. **Post-processing**: after the model generates a response, you could try applying some rules or filters (e.g., with the help of keyword lists, regular expressions, or even another ML model) to remove or alter parts that are off-topic or inappropriate.
2. **Iterative Refinement**: you could try taking the model's output and use it in a subsequent prompt to which you could add what to include/exclude. This should allow the model to "self-correct"/stay on topic over several iterations.
3. **Model Improvement**: if you have access to the model, you could fine-tune it by training on a specific dataset (e.g., which is based on your context). Or you could provide feedback to its answers (good/bad) and fine-tune it in that way (though that would require a lot of interaction, with many users).

But **prompt-engineering** is the easiest option, you may want to check if your LLM accepts negative prompts, and, if it does, create a big list of topics you don't want the LLM to consider when answering.

**Comment by MAVRICK-1** [comment]
Reducing out-of-context queries to the LLM is a challenge that many developers and researchers face when using generative AI models. Out-of-context queries are those that do not match the domain, task, or format that the LLM is trained or fine-tuned for. For example, asking a LLM that is designed for text summarization to write a poem, or asking a LLM that is specialized for medical questions to answer trivia. Out-of-context queries can result in inaccurate, irrelevant, or nonsensical outputs from the LLM.

There are different ways to reduce out-of-context queries to the LLM, depending on the type and purpose of the application. Some of the common methods are:

- **Optimizing the prompt**: The prompt is the input that is given to the LLM to generate an output. A well-designed prompt can provide clear instructions, expectations, and constraints for the LLM. For example, using keywords, prefixes, suffixes, templates, examples, and feedback signals can help guide the LLM to produce relevant and coherent outputs. A poorly-designed prompt can confuse or mislead the LLM, resulting in out-of-context outputs¹⁴.
- **Optimizing the context**: The context is the additional information that is given to the LLM along with the prompt. The context can include documents, facts, knowledge bases, or previous interactions that are related to the domain or task of the application. A well-chosen context can provide useful background and references for the LLM to generate accurate and consistent outputs. A poorly-chosen context can introduce noise or contradictions for the LLM, resulting in out-of-context outputs²³.
- **Optimizing the model**: The model is the LLM itself, which can be pre-trained or fine-tuned on a large corpus of text data. A well-trained model can capture the general patterns and semantics of natural language, as well as the specific features and nuances of a particular domain or task. A poorly-trained model can have gaps or biases in its knowledge or skills, resulting in out-of-context outputs¹².

These are some of the methods that can help reduce out-of-context queries to the LLM. However, there is no guarantee that these methods will work perfectly for every application or scenario. Therefore, it is important to test and evaluate the LLM outputs regularly and rigorously, and apply appropriate quality control and error handling mechanisms¹².

(1) Reducing out-of-context queries to the LLM #67504 - GitHub. https://github.com/orgs/community/discussions/67504.
(2) How to reduce the costs of using ChatGPT and GPT-4. https://bdtechtalks.com/2023/06/12/reduce-costs-of-chatgpt-gpt-4/.
(3) How to Get Better Outputs from Your Large Language Model. https://developer.nvidia.com/blog/how-to-get-better-outputs-from-your-large-language-model/.
(4) Optimizing Large Language Models for Cost Efficiency - Vantage. https://www.vantage.sh/blog/optimize-large-language-model-costs.
(5) LLM's answering out of context ( trained on user data). https://stackoverflow.com/questions/76751579/llms-answering-out-of-context-trained-on-user-data.

**Comment by graphicmist** [comment]
GPT is very good in response and its output can easily be controlled with prompt engineering for answering within a specific context and knowledge-base. However this technique is not very effective with llama, mistral and other open source models.
